### ISSUE TITLE
[explicit-module-boundary-types] New option to allow unspecified return type if there are overload signatures

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#4586]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

[Playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=KYDwDg9gTgLgBAMwK4DsDGMCWEWIhACgEMAuOAZxikxQHMBKMy6ugbgChRJZFUNtcCfMTIokAWwBGwKIzhipMjl2jxk6LDjyFScVAGsUEAO4p6cAN7s4NuFGAwkUXEQ4BfdkA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9HAB4z5JrYC2A9gCYCuucWARvcwYwIYooWaPAQgAXCDgoU9FCAC+QA&tsConfig=N4XyA)

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": "error"
  }
}
```

```TS
export function foo(a: string): string;
export function foo(a: number): number;
export function foo(a: unknown) {
  return a;
}
```

**Expected Result**

I'd like a new option that makes this code pass, since the implementation signature is not visible to the outside, while the _boundary_, defined by the overload signatures, is well-typed.

**Actual Result**

This errors, even with all the `allow*` options turned on.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.1` |
| `@typescript-eslint/parser`        | `5.12.1` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.9.0` |
| `node`                             | `17.5.0` |

**Comment by bradzacher** [comment]
I don't think that this is a good idea because it can lead you to writing a bad implementation.

Your example shows one of the major type-safety holes - an implementation that returns `unknown` is considered compatible with all strictly typed signatures.

But we know that isn't safe or correct because `unknown` is not compatible. For example

```ts
export function foo(a: string): string;
export function foo(a: number): number;
export function foo(a: unknown) {
  return (true as unknown);
}
```
[Repl](https://www.typescriptlang.org/play?#code/AQUwHgDg9gTgLsAZgVwHYGM4EsqqVKACgEMAuYAZzhi1QHMBKcqm+gbgChRJYEUNsufETLBUyALYAjEDCZjJMmJw7ho8JGkw48iAiXJoA1qigB3VA2ABvLsBgg4yGHkLVkIYMQrBjpiwycAL4cdnpQHAD0kcAAegD8QA)

[With an explicit annotation you know the types are going to be more correct within the implementation](https://www.typescriptlang.org/play?#code/AQUwHgDg9gTgLsAZgVwHYGM4EsqqVKACgEMAuYAZzhi1QHMBKcqm+gbgChRJYEUNsufETLBUyALYAjEDCZjJMmJw7ho8JGkw48iAiXJoA1qigB3VPJa06wAD4Lps4AG8uwGCDjIYeQtWQQYGIKYGNTCwZOAF8Odz0oDgB6JOAAPQB+IA), and it's less likely you're going to make a mistake.
There's still the option to do something dodgy like an explicit unknown return type - but at least that is obvious that the contract is dodgy.

The entire point of the rule is to enforce strict contracts at the boundary, so I don't think ignoring the return type here aligns with the rule itself.

**Comment by Josh-Cena** [comment]
I respectfully disagree. This rule is about the _contract_, not about preventing implementation mistakes. Is the module boundary explicit?—Yes, fully explicit. Is the implementation checked against the module boundary?—Yes, at least with at least some non-trivial effort. I drew that conclusion from the `allow*` options that we have today, which seem to err towards ergonomics than fool-proofness, especially the `allowDirectConstAssertionInArrowFunctions`. For example, if I have a function: `const a = () => ({ foo: 1, bar: 2 } as const);`, I can easily get rid of `bar`, or change it to `bar: 'baz'`, which breaks my contract but doesn't get reported.

If you argue that overload signatures are essentially downcasts at the boundary, then I'd say that the entire thing about overload signatures usually involves developer discretion by carefully exposing a subset of what their function can do to the end-user. For example, I may also expose a wrong subset:

```ts
export function foo(a: string): string;
export function foo(a: number): number;
export function foo(a: string | number): string | number {
  if (typeof a === 'string') {
    return Number(a);
  }
  return String(a);
}
```
